pages/doc.js

```diff
@@ -25,216 +24,149 @@ import { media } from '../src/styles'
-                pageNotFound: false
-              },
-              () => this.updateScroll(!isFirstPage && isPageChanged)
-            )
-          })
-        })
-        .catch(() => {
-          window.location.reload()
-        })
-    }
-  }
+        <Side isOpen={isMenuOpen}>
+          {isSearchAvaible && (
+            <SearchArea>
+              <SearchForm />
+            </SearchArea>
+          )}
+
+          <SidebarMenu
+            sidebar={sidebar}
+            currentPath={path}
+            id={SIDEBAR_MENU}
+            onClick={toggleMenu}
+          />
+        </Side>
+
+        <Markdown
+          markdown={markdown}
+          githubLink={githubLink}
+          prev={prev}
+          next={next}
+          tutorials={tutorials}
+        />
+
+        <RightPanel
+          headings={headings}
+          githubLink={githubLink}
+          tutorials={tutorials}
+        />
+      </Container>
+    </Page>
+  )
+}
 
-  updateScroll(isPageChanged) {
-    const { hash } = window.location
+Documentation.getInitialProps = async ({ asPath, req }) => {
+  const item = getItemByPath(asPath)
 
-    if (isPageChanged) {
-      this.setState({ isSmoothScrollEnabled: false }, () => {
-        this.scrollTop()
-        this.setState({ isSmoothScrollEnabled: true }, () => {
-          if (hash) this.scrollToLink(hash)
-        })
-      })
-    } else if (hash) {
-      this.scrollToLink(hash)
+  if (!item) {
+    return {
+      errorCode: 404
     }
   }
 
-  parseHeadings = text => {
-    const headingRegex = /\n(## \s*)(.*)/g
-    const matches = []
-    let match
-    do {
-      match = headingRegex.exec(text)
-      if (match)
-        matches.push({
-          text: match[2],
-          slug: kebabCase(match[2])
-        })
-    } while (match)
-
-    return matches
-  }
+  const host = req ? req.headers['host'] : window.location.host
+  const protocol = host.indexOf('localhost') > -1 ? 'http:' : 'https:'
 
-  scrollToLink = hash => {
-    const element = document.getElementById(hash.replace(/^#/, ''))
+  try {
+    const res = await fetch(`${protocol}//${host}${item.source}`)
 
-    if (element) {
-      const headerHeight = document.getElementById(HEADER).offsetHeight
-      const elementBoundary = element.getBoundingClientRect()
-      const rootElement = document.getElementById(ROOT_ELEMENT)
-      rootElement.scroll({ top: elementBoundary.top - headerHeight })
+    if (res.status !== 200) {
+      return {
+        errorCode: res.status
+      }
     }
-  }
 
-  scrollTop = () => {
-    const rootElement = document.getElementById(ROOT_ELEMENT)
-    if (rootElement) {
-      rootElement.scrollTop = 0
-    }
-  }
+    const text = await res.text()
 
-  toggleMenu = () => {
-    this.setState(prevState => ({
-      isMenuOpen: !prevState.isMenuOpen
-    }))
+    return {
+      item: item,
+      headings: parseHeadings(text),
+      markdown: text
+    }
+  } catch {
+    window.location.reload()
```

Comment:
Can't this go into an infinite loop? (Blocked by the browser probably.)

> Continuation of https://github.com/iterative/dvc.org/pull/834#pullrequestreview-327869118

Reply:
If page loads again then yes, but this error usually means that the internet is down and we will just show browser error page there. Maybe there are cases then this error is thrown in other situations, but I don't know about them.